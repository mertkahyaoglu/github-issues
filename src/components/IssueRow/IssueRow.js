import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import ExclamationIcon from 'react-icons/lib/fa/exclamation-circle';
import CommentIcon from 'react-icons/lib/fa/comment';
import Status from 'components/Status';
import Label from 'components/Label';
import Assignee from 'components/Assignee';
import { timeDifference } from 'utils/helpers';

export default class IssueRow extends Component {
  static propTypes = {
    data: PropTypes.object.isRequired,
  };

  renderLabels = labels =>
    labels.map(label => <Label key={label.id} label={label} />);

  renderAssignees = assignees =>
    assignees.map(assignee => (
      <Assignee key={assignee.id} assignee={assignee} />
    ));

  render() {
    const { data } = this.props;
    return (
      <Row>
        <Status
          status={data.state === 'open' ? 'success' : 'danger'}
          icon={ExclamationIcon}
        />
        <IssueInfo>
          <TitleWrapper>
            <Title href={data.html_url} target="_blank">
              {data.title}
            </Title>
            <Labels>{this.renderLabels(data.labels)}</Labels>
            <Assignees>{this.renderAssignees(data.assignees)}</Assignees>
            {data.comments > 0 && (
              <CommentsLink href={data.html_url} target="_blank">
                <Status icon={CommentIcon}>{data.comments}</Status>
              </CommentsLink>
            )}
          </TitleWrapper>
          <Meta>
            {`#${data.number} opend on ${timeDifference(
              +new Date(),
              +new Date(data.created_at)
            )} by`}{' '}
            <UserLink href={data.user.html_url} target="_blank">
              {data.user.login}
            </UserLink>
          </Meta>
        </IssueInfo>
      </Row>
    );
  }
}

const Row = styled.div`
  display: flex;
  align-items: flex-start;
  padding: 13px;
  background-color: #fff;
  border-bottom: 1px solid #e1e4e8;
  &:hover {
    background-color: #f6f8fa;
  }
`;

const IssueInfo = styled.div`
  width: 100%;
`;

const TitleWrapper = styled.div`
  display: flex;
  align-items: center;
`;

const Title = styled.a`
  color: #24292e;
  font-weight: 600;
`;

const UserLink = styled.a`
  color: #586069;
`;

const CommentsLink = styled.a`
  color: #586069;
  font-size: 12px;
`;

const Labels = styled.div`
  margin-left: 4px;
`;

const Assignees = styled.div`
  margin-left: auto;
  margin-right: 10px;
`;

const Meta = styled.div`
  color: #586069;
  font-size: 12px;
  margin-top: 8px;
`;
