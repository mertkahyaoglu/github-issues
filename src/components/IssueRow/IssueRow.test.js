import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import IssueRow from './IssueRow';
import TestProviders from 'utils/TestProviders';

describe('component/IssueRow', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <TestProviders>
        <IssueRow data={data} />
      </TestProviders>,
      div
    );
  });

  it('renders markup correctly', () => {
    const tree = renderer
      .create(
        <TestProviders>
          <IssueRow data={data} />
        </TestProviders>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const data = {
  url: 'https://api.github.com/repos/jekyll/jekyll-admin/issues/456',
  repository_url: 'https://api.github.com/repos/jekyll/jekyll-admin',
  labels_url:
    'https://api.github.com/repos/jekyll/jekyll-admin/issues/456/labels{/name}',
  comments_url:
    'https://api.github.com/repos/jekyll/jekyll-admin/issues/456/comments',
  events_url:
    'https://api.github.com/repos/jekyll/jekyll-admin/issues/456/events',
  html_url: 'https://github.com/jekyll/jekyll-admin/issues/456',
  id: 307287639,
  number: 456,
  title: 'Security issues',
  user: {
    login: 'theunluckyone1',
    id: 37171438,
    avatar_url: 'https://avatars2.githubusercontent.com/u/37171438?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/theunluckyone1',
    html_url: 'https://github.com/theunluckyone1',
    followers_url: 'https://api.github.com/users/theunluckyone1/followers',
    following_url:
      'https://api.github.com/users/theunluckyone1/following{/other_user}',
    gists_url: 'https://api.github.com/users/theunluckyone1/gists{/gist_id}',
    starred_url:
      'https://api.github.com/users/theunluckyone1/starred{/owner}{/repo}',
    subscriptions_url:
      'https://api.github.com/users/theunluckyone1/subscriptions',
    organizations_url: 'https://api.github.com/users/theunluckyone1/orgs',
    repos_url: 'https://api.github.com/users/theunluckyone1/repos',
    events_url: 'https://api.github.com/users/theunluckyone1/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/theunluckyone1/received_events',
    type: 'User',
    site_admin: false,
  },
  labels: [],
  state: 'open',
  locked: false,
  assignee: null,
  assignees: [],
  milestone: null,
  comments: 0,
  created_at: '2018-03-21T15:13:00Z',
  updated_at: '2018-03-21T15:13:00Z',
  closed_at: null,
  author_association: 'NONE',
};
