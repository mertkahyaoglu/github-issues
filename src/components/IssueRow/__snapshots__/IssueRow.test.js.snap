// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/IssueRow renders markup correctly 1`] = `
<div
  className="sc-ifAKCX lHalS"
>
  <div
    className="sc-bdVaJa hiZeDy"
  >
    <svg
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "marginRight": "0.5em",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 40 40"
      width="1em"
    >
      <g>
        <path
          d="m20.1 2.9q4.7 0 8.6 2.3t6.3 6.2 2.3 8.6-2.3 8.6-6.3 6.2-8.6 2.3-8.6-2.3-6.2-6.2-2.3-8.6 2.3-8.6 6.2-6.2 8.6-2.3z m2.9 27.8v-4.2q0-0.4-0.2-0.6t-0.5-0.2h-4.3q-0.3 0-0.5 0.2t-0.2 0.5v4.3q0 0.3 0.2 0.5t0.5 0.2h4.3q0.3 0 0.5-0.2t0.2-0.5z m0-7.7l0.4-13.8q0-0.3-0.3-0.5-0.2-0.1-0.5-0.1h-4.9q-0.3 0-0.5 0.1-0.3 0.2-0.3 0.4l0.4 13.9q0 0.2 0.2 0.4t0.6 0.2h4.1q0.3 0 0.5-0.2t0.3-0.4z"
        />
      </g>
    </svg>
  </div>
  <div
    className="sc-EHOje YINzU"
  >
    <div
      className="sc-bZQynM jdkaLI"
    >
      <a
        className="sc-gzVnrw dUTaIu"
        href="https://github.com/jekyll/jekyll-admin/issues/456"
        target="_blank"
      >
        Security issues
      </a>
      <div
        className="sc-iwsKbI bshbDz"
      />
      <div
        className="sc-gZMcBi hWXGrZ"
      />
    </div>
    <div
      className="sc-gqjmRU kNPpqd"
    >
      #456 opend on 2018-03-21T15:13:00Z by
       
      <a
        className="sc-htoDjs gjQERW"
        href="https://github.com/theunluckyone1"
        target="_blank"
      >
        theunluckyone1
      </a>
    </div>
  </div>
</div>
`;
